controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,470 @@
+
+	// Set the start time and add the pods to the in progress list
+	patch := client.MergeFrom(task.DeepCopy())
+	task.Status.StartTime = metav1.Now()
+	for _, pod := range pods {
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	// Invoke the purge operation on all pods, in the background
+	go func() {
+		wg := sync.WaitGroup{}
+
+		// Mutex to prevent concurrent updates to the backup.Status object
+		backupMutex := sync.Mutex{}
+		patch := client.MergeFrom(task.DeepCopy())
+
+		for _, p := range pods {
+			pod := p
+			wg.Add(1)
+			go func() {
+				logger.Info("starting purge", "CassandraPod", pod.Name)
+				succeeded := false
+				var response *medusa.PurgeBackupsResponse
+				response, purgeErr := doPurge(ctx, &pod, r.ClientFactory)
+				if purgeErr == nil {
+					logger.Info("finished purge", "CassandraPod", pod.Name)
+					succeeded = true
+				} else {
+					logger.Error(purgeErr, "purge failed", "CassandraPod", pod.Name)
+				}
+				backupMutex.Lock()
+				defer backupMutex.Unlock()
+				defer wg.Done()
+				task.Status.InProgress = utils.RemoveValue(task.Status.InProgress, pod.Name)
+				if succeeded && response != nil {
+					purgeResponse := medusav1alpha1.TaskResult{
+						PodName:                   pod.Name,
+						NbBackupsPurged:           int(response.NbBackupsPurged),
+						NbObjectsPurged:           int(response.NbObjectsPurged),
+						TotalPurgedSize:           int(response.TotalPurgedSize),
+						TotalObjectsWithinGcGrace: int(response.TotalObjectsWithinGcGrace),
+					}
+					task.Status.Finished = append(task.Status.Finished, purgeResponse)
+				} else {
+					task.Status.Failed = append(task.Status.Failed, pod.Name)
+				}
+			}()
+		}
+		wg.Wait()
+		logger.Info("finished task operations")
+		if err := r.Status().Patch(context.Background(), task, patch); err != nil {
+			logger.Error(err, "failed to patch status", "MedusaTask", fmt.Sprintf("%s/%s", task.Spec.Operation, task.Namespace))
+		}
+	}()
+
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+func (r *MedusaTaskReconciler) prepareRestoreOperation(ctx context.Context, task *medusav1alpha1.MedusaTask, pods []corev1.Pod, logger logr.Logger) (reconcile.Result, error) {
+	logger.Info("Starting prepare restore operations")
+
+	// Set the start time and add the pods to the in progress list
+	patch := client.MergeFrom(task.DeepCopy())
+	task.Status.StartTime = metav1.Now()
+	for _, pod := range pods {
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
```

Comment:
No need to set `RequeueAfter` when returning a non-nil error.

Reply:
👍